### ISSUE TITLE
Pet Set in beta

### ISSUE THREAD

**Issue by bprashanth** [kubernetes/kubernetes#28718]
Important petset issues in no particular order:
1. Local storage: https://github.com/kubernetes/kubernetes/issues/7562
2. Upgrade: https://github.com/kubernetes/kubernetes/issues/28706
3. Richer events: https://github.com/kubernetes/kubernetes/issues/24957
4. Public ips: https://github.com/kubernetes/kubernetes/issues/28660
5. Leader election and failover: https://github.com/kubernetes/kubernetes/issues/1542, https://github.com/kubernetes/kubernetes/issues/28658
6. Non database examples: eg Kafka on-top of zookeeper petset (https://github.com/kubernetes/kubernetes/issues/23794)
7. Petset fencing: https://github.com/kubernetes/kubernetes/issues/31762

IMO the following features would make PetSet more usable and should gate beta:
1. Local storage in some form, so people can deploy databases without cloud provider help
2. Basic rolling-update style upgrades as described in (2) 
3. 1 non-database example like Kafka (which might bring more feature requests with it)
4. Fencing?

Richer events would be really nice but I don't think it blocks beta unless we need it for 2 or 3. Lack of public ips will probably block WAN deployments. 

@smarterclayton @kubernetes/sig-apps anything else?

**Comment by chrislovecnm** [comment]
Sticky ips ;) I don't think Cassandra and other tech is going to play nice without ip addresses. Java DNS is hosed, and products that have been around awhile are still using ip addresses. I think Elastic does as well. Need to check.

**Comment by chrislovecnm** [comment]
Btw unless someone gets to it, I will ask if my client, or I can contribute Kafka on pet set. We are going to be using it in production soon.

**Comment by bprashanth** [comment]
> Btw unless someone gets to it, I will ask if my client, or I can contribute Kafka on pet set

Go for it. Keep communicating hurdles you run into. 

>Sticky ips ;) I don't think Cassandra and other tech is going to play nice without ip addresses. Java DNS is hosed

Duly noted. grr.

**Comment by chrislovecnm** [comment]
@bprasanth you need an issue for sticky ip addresses?

**Comment by smarterclayton** [comment]
We've had lots of very positive results with low TTL DNS in Java with JBoss
and other portfolio products - I'm not that skeptical it can't be fixed.

**Comment by smarterclayton** [comment]
I'd put reducing the complexity required to make a "correct" petset at the
top of the list - I'd even be willing to postpone other more sophisticated
operationalization things in favor of that.  If someone has to be an expert
at distributed systems to build a correct PetSet we'll probably not succeed.

**Comment by chrislovecnm** [comment]
@smarterclayton it is about if it can be fixed, it is about how distributed systems have implemented it. I need to check on elastic, but I am 90% certain that Cassandra is not going to fix it. Going to file an issue with them anyways.

**Comment by bprashanth** [comment]
> I'd put reducing the complexity required to make a "correct" petset at the
top of the list - I'd even be willing to postpone other more sophisticated
operationalization things in favor of that.  If someone has to be an expert
at distributed systems to build a correct PetSet we'll probably not succeed.

This is less of a problem for modern databases like etcd, but it is a problem by and large. 

With a better event model the system can pass down observations about the current state of the cluster to individual members. The exact events will probably depend on the type of PetSet. Obviously we can't have a type=Cassandra, and say, "tell me when the seeds are down", but we can say something like "index 0,1" are always seeds, send all members of the set a role based event if they fail a probe. 
 
This requires bidirectional event/response infrastructure. I *think* the easiest way to do this is with a sidecar (might need shared pid namespaces). For example we could add the following fields to petset

```yaml
checkIndices: 0
jsonProbe:
  httpGet:
      path: /
```

Where the json probe returns whatever metadata a member needs to elect itself into that role, if any (eg: transaction id). All the sidecars would then heartbeat the probe, and race for the position on failure. The winner writes their index into checkIndices and the process continues. 

Did you mean something like this? Getting it right is *tricky* (shortcomings of a http health check, flapping leaders, leader resurrection etc), but it's probably what a lot of people are doing themselves.

**Comment by magicwang-cn** [comment]
@kubernetes/huawei

**Comment by haizaar** [comment]
ElasticSearch uses dedicated [k8s](https://github.com/fabric8io/elasticsearch-cloud-kubernetes) plugin for node discovery. We are using it now in k8s 1.2 and we don't need sticky IPs for that.

**Comment by haizaar** [comment]
No :(

We are ES-only shop.

**Comment by damoon** [comment]
Ceph Monitors and etcd need sticky IPs as far as I know. Or at least fixed dns names. Currently i use a service per 1 Pod deployment with hostpath volume and node Selector.

**Comment by bprashanth** [comment]
Petset already gives you fixed dns names http://kubernetes.io/docs/user-guide/petset/#network-identity. If you have some working form of ceph with petset I'd be glad to help work out the kinks/understand what we need to grow petset. I believe theres a prototype etcd in the works: https://github.com/kubernetes/contrib/pull/1295

**Comment by chrislovecnm** [comment]
@damoon I am looking for a use case that requires IP addresses to be sticky.  Otherwise, I will not ask for it.  I think we are coming up dry.

**Comment by bprashanth** [comment]
Most pets are not going to run amock in the 1000s. For those that do, you can write a custom autoscaler. Ideally we would integrate with HPA but unless someone has hard use cases for a broad category of pets that need autoscaling I'm tempted to punt on that for beta. 

we should add a flag that allows auto-GC of the pvc when a pet is killed.

**Comment by chrislovecnm** [comment]
@bprashanth I kinda don't agree.  These are persistent data stores, and having autoscaling based on memory or cpu would be HUGE.

**Comment by bprashanth** [comment]
HPA is already aware of how to scale RCs, teaching it to bump up the replica count on a petset can happen without any additional features or api changes, if someone does the plumbing. @kubernetes/autoscaling 

autoscaling cross zone/region is a different story.  For that we need WAN deployment prototypes.

**Comment by smarterclayton** [comment]
Not probably - are, and having to rediscover every failure.  Distributed
systems need training wheels, and Kubernetes needs to be the tricycle.

**Comment by damoon** [comment]
ceph: http://docs.ceph.com/docs/hammer/rados/operations/add-or-rm-mons/#changing-a-monitor-s-ip-address ceph mons NEED static ips

etcd: fixed dns names are good enough. but it seems, a dns request from a pod targeting itself fails for some reason. seems kubedns tries to prevent services to call themself recursively.

**Comment by bprashanth** [comment]
Lets maintain a list of apps that needs ip on the sticky ips bug (https://github.com/kubernetes/kubernetes/issues/28969) so we can at least start by cautioning users in documentation. Currently you can create a Service pet pet for sticky ips. 

@damoon can you comment on the deployment style you'd use for ceph. It sounds like one might want to run ceph as a daemon set and only the monitors as a petset, if petset had sticky ips?

**Comment by magicwang-cn** [comment]
i think the petset can join the concept of (`Role`).  Which represents a component of an app, every role has its (`Replicas, PodSpec, etc.`)
for example:

 - mongodb is the composition of config server,  router and shard, these three components can  be         three roles. 
- etcd has just one role

what's more, we can supply a common image(like peer) to find every role's identity information(like ip-address, port, hostname, etc.) by init-containers, and store these information to a specific config file, so users can get these information from the config file, start their apps by themselves.

**Comment by damoon** [comment]
i have 2 working ceph setups.
one uses manually configured kubernetes external containers via weave with static ips for the monitors. this was/is production for quite some time.
the 2nd uses a daemonset for the datanodes (osd) and one deployment + service per monitor with static ips for the services. each services points to a single monitor and each pod is fixed to a kubelet-node. telling the ceph config the service ips are the monitors seems to work fine. this is only tested locally with vagrant and is not battle proven.

**Comment by leseb** [comment]
@bprashanth basically when it comes to ceph monitor nodes, once they get deployed their IP address can not be changed. That's why we need sticky IP addresses. Other components can have their IP changed.

**Comment by bprashanth** [comment]
Summoning @erictune for views on the "make distributed systems easier through petset and kube" agenda

**Comment by smarterclayton** [comment]
I'm ok with the list in general.  I'd like to get to a point where we identify all the "common needs".  I'll update the proposal this week with the current state and this list, as well as other unresolved questions.

**Comment by magicwang-cn** [comment]
@zhulinhong

**Comment by chrislovecnm** [comment]
One thing I noticed is that we are not adding scaling via `kubectl`.  For example:

```sh

$ kubectl scale petset cassandra --replicas=3

```

This would be a nice to have.

**Comment by bprashanth** [comment]
Yeah im writing a scaler/reaper right now

**Comment by chrislovecnm** [comment]
Awesome ;)

**Comment by smarterclayton** [comment]
Proposal is updated to capture these goals and also to set some GA goals.

**Comment by Cloven** [comment]
As a non-database example, erlang nodes identify themselves and each other either by 'nodename@shorthostname' or 'nodename@canonical.host.name' (http://erlang.org/doc/reference_manual/distributed.html).  

When they're configured to do so, they form TCP mesh connections once they are aware of other nodes.  They do this using the erlang port mapper daemon (epmd) (http://erlang.org/doc/man/epmd.html), which maps nodenames to the nodes' reported IP addresses at time of discovery.  

Erlang programs then expect to be able to send messages, run functions, etc., on other nodes using the bare name of the node (e.g. {shell, felix@testnode} ! boop.), and can also take references to remote processes, pass them around, and confidently expect to be able to dereference them later.

Clustered RabbitMQ is probably the most canonical and widespread example dependent on this facility, but I've also seen a fair amount of custom software relying on the same identity-to-ip correlation assumptions.

**Comment by bprashanth** [comment]
@Cloven if you happen to try rabbit with petset please report the hurdles you ran into. We're lacking an example workqueue, restricted ourselves to dbs till now. From your description it sounds like epmd will only resolve a dns name one time (vs respecting ttl and re-resolving)?

**Comment by analytik** [comment]
Hi,
Sorry to derail this with a lateral comment, but regarding auto-scaling:

Scaling RabbitMQ would make sense to scale for perishable (or persistent, non-replicated) messages. That would be one use case that would benefit from auto-scaling.

However, persistent RabbitMQ topics, and of course any kind of persistent storage would get slower if their replica count got increased. For example with RethinkDB, it's easy to (auto)scale up and join a cluster, but not very useful without a custom startup script that would reshard/replicate tables that have high read loads. And such a script might be too clunky to run on startup, so to me, HPA on PetSets would make sense only if it's coupled with an option to have a scaling event trigger a custom script (e.g. rebalance and/or replicate and/or shard tables up or down based on current read/write load).

**Comment by smarterclayton** [comment]
I think reshard / replicate falls under the "general cluster maintenance"
responsibilities that a babysitter / controller process would have, but
since it's somewhat specialized it's also in the category of "could just be
an admin problem".  The thing we want to ensure is possible is that an
admin *can* write that script (to reshard / resize) dynamically if they
wanted and not have to think too hard about how to do it correctly.

**Comment by Cloven** [comment]
@bprashanth erlang's epmd doesn't resolve names or use dns; nodes name themselves from their hostname on startup, and then pass their name and their ip address to epmd.  If it helps, erlang's essentially a tiny embedded operating system built for use in telephone switches, where the expectation of stable ip, name, and topology, and optionality of such deluxe niceties as dns service, is the norm.

I currently use GKE so can't test RabbitMQ (or clustered erlang generally) on PetSets as easily as I'd like, but since the current spec doesn't seem to allow for ip address stability, I suspect it'll have weird edgy issues at a minimum.

More generally, and I confess to being personally guilty of this, there are many distributed systems that store IP addresses, often in order to avoid the cost of future DNS lookups, without storing TTL and faithfully invalidating the IP address when the TTL time arrives.  If PetSet had the ability to stickily keep a hostname with an ip address, then the problem of a node or cluster member being killed and rescheduled could look like a standard, well understood computer crash to all those existing algorithms.  In the absence of that, then k8 would be creating a new behavior unexpected to those algorithms in which it could appear that a computer is killed and a new impostor computer tries to join the cluster using the old computer's identifier but at a different location.  This is a rare case in metal clusters in my experience, rare enough that lazy algorithm authors leave detection and resolution to operator intervention, but could be significantly more common in k8.

**Comment by smarterclayton** [comment]
We definitely want to do what we can to enable those people who are able to
return to trusting DNS (I think DNS in general has improved by leaps and
bounds outside of the enterprise in the intervening years), but we'll have
to have some fallback for those who are dependent on not trusting it until
a few years down the road when they can relax that dependency.  HAProxy is
a good example of having a multi-year effort to go back to allowing DNS to
be trusted (with endpoint listing possible via DNS).

We need some serious design time on what sticky IPs might look like and how
to make them part of the virtual infrastructure.  I do think one wrinkle is
that even with sticky IPs there are very real transitions that happen under
the covers, and the implications of that transition can still result in two
peers on the network having the same IP at a point in time (and propagation
delays).  We definitely need to understand the other challenges that could
arise in those transitions beyond just networking.

**Comment by chrislovecnm** [comment]
@smarterclayton - should we hijack a sig apps session for design? Like to get the right people on a video call and bang out at least a rough draft.

**Comment by smarterclayton** [comment]
Yeah, maybe the week after code slush.

**Comment by Cloven** [comment]
First, is anything happening on this?  I'd be interested in listening in to any publicly-accessible call if it hasn't happened yet, this topic is super intriguing.

Second, @smarterclayton writes:
>  I do think one wrinkle is
that even with sticky IPs there are very real transitions that happen under
the covers, and the implications of that transition can still result in two
peers on the network having the same IP at a point in time (and propagation
delays).  We definitely need to understand the other challenges that could
arise in those transitions beyond just networking.

Where do I go to learn more about the nature of these transitions?  Once an IP has been assigned to any entity, I'd expect herculean efforts to take place to not permit any other entity on the network to believe it also has the same IP.  There's a lot in k8s that seems to rely on that assumption -- IP reservation is read through etcd for services, for instance, or service IPs being kept in immutable environment variables.  The only thing I could imagine is if you're trying to use something akin to corosync/pacemaker/heartbeat/etc. to provide HA, but errors in the failover window are a decently rare case.

**Comment by smarterclayton** [comment]
A petset does not guarantee only a single pod identifying as any member
exists (you can have 2 running processes on different nodes that think they
are mysql-0).  It only guarantees that you can determine *which one* should
be included in the official config.  Any network magic for external sticky
IPs would still have to deal with this fundamental split - if you do that
by programming edge routers, for instance, there can be a period of time a
client can reach both IPs.  Distributing an IP address routing change to
multiple machines is still a consensus problem and you can't guarantee that
they all get it at the same time.

**Comment by Cloven** [comment]
> A petset does not guarantee only a single pod identifying as any member
exists (you can have 2 running processes on different nodes that think they
are mysql-0). 

That looks to be contradicted by the Pet Set documentation (http://kubernetes.io/docs/user-guide/petset/):

>  Each Pet has a deterministic name - PetSetName-Ordinal, and a unique identity. Each Pet has at most one pod, and each Pet Set has at most one Pet with a given identity.

And, I'd find it pretty surprising.  Can you describe a scenario where you think that should or would happen?  How would the impostor mysql-0 behave?

**Comment by smarterclayton** [comment]
1. Create pod mysql-0, scheduled to node-0
2. Network partition between node-0 and the cluster
3. Node controller detects node-0 is down, evacuates node by deleting pods
4. pod mysql-0 reaches maximum grace period without node coming online, deleted by node controller
5. pet set controller creates mysql-0 again
6. Network partition heals between node-0 and rest of cluster

mysql-0 was always available, and could have been serving traffic (depending on what views other clients have).

PetSets do their best to provide a consistent mapping, but just because a pod has been deleted doesn't mean the process doesn't still exist.  You can increase the partition tolerance time indefinitely by setting higher pod graceful deletion periods, but if you choose availability over consistency you will have two members running.

The guarantee petset provides is that the other cluster members with access to the master can observe that the initial mysql-0 was deleted, and use the new mysql-0 in their quorum.

**Comment by Cloven** [comment]
oh yeah, for sure you wouldn't ever do 3b, 4 or 5 with a pet, would you?  That'd be treating a pet like cattle and breaking the key underpinning abstractions of Pet Sets.  Like operationally, what are you supposed to do when two postgres masters have both taken traffic and your data is not stored as CRDTs?  Or when two paxos nodes come up thinking they have the same identifier and voting accordingly?

**Comment by smarterclayton** [comment]
3, 4, 5 are all the default behavior of Kube.  There is no mechanism today to avoid that, other than setting arbitrarily long grace periods.

The point of pet set is to create reliable abstractions on top of unreliable entities.  The potential for split brain has to be managed as the cluster is reconfigured - for instance, mysql-0 prime cannot just act as if it is primary - it must engage in a process to determine whether it is indeed the primary.  Access to a shared disk can assist with that for some storage mediums (the disk operates as a lock for GCE/AWS), but the point of this issue is to identify and build in mechanisms that allow cluster operators to correctly build clustered software on top of pet sets safely.

The fact that you could be confused by the documentation into believing that split brain is impossible is a good example of some of the steps that need to happen before pet sets go beta.

**Comment by smarterclayton** [comment]
Moving to 1.5 - I've added this to the list of proposed work for sig-apps in 1.5.  I'm somewhat skeptical at this point that we can move to beta in 1.5, but we should definitely make progress towards the things that allow beta.

**Comment by bprashanth** [comment]
I think there's a blurry line between what the system should handle vs what would be better handled directly by the app, and that's the challenge we're facing right now. Using a pd as a lock is a clunky solution, but it allows us to move past this particular problem. The same can be said about sticky ips, public identities or leader election support in legacy dbs. 

There's a list of things that we can document as known caveats, build *more* production ready prototypes, and then circle back to graduation requirements. The first part needs to be a community effort IMO, discussing this at next community meeting/sig-apps SGTM.

**Comment by Cloven** [comment]
> for instance, mysql-0 prime cannot just act as if it is primary - it must engage in a process to determine whether it is indeed the primary. 

Even if there were a system of leader election in mysqld through which all queries ran (!!), how do A, B, C, A and B again form a valid quorum?

As documented, I think Pet Sets work swimmingly; they provide a stable guarantee of identity and a stable name and a consistent zero or one quantity, with all that implies.  It follows the price of entry is that cracked Pet Sets require human intervention, like a failed Deployment.  This other idea of Pet Sets didn't even occur to me because the dangers are so tall and the guarantees so faint.

**Comment by bprashanth** [comment]
At the same time I don't think it's on kube to reinvent paxos. The best we can do is provide a framework, where if you assume an HA control plane, you minimize the probability of catastrophic partitions.  The app needs to layer a consistency framework on top. Common methods involve taking a lease in etcd (see frameworks like patroni: https://github.com/zalando/patroni), blocking in the init container on a lock so the second mysql is simply not allowed to start (blocking on a cloudprovider pv achieves similar effect), or running a sidecar hearbeater like redis sentinel. I don't think this is a new problem in distributed computing.

Of course having a petset annotation serve as a hint to the nodecontroller is easy, and exactly the sort of thing we can probably ignore for now knowing there's an escape hatch, but the space is rife with such race conditions.

**Comment by Cloven** [comment]
Patroni is pretty interesting, thanks for the link!   And you make the good point that there are valid, if complex, workarounds in the case where you really do want to extend the availability concern into Pet Sets.  Food for thought.

**Comment by smarterclayton** [comment]
Yeah - we have to walk a very fine line between providing tools that make "correct" and "safe" distributed systems possible.  I.e., a moderately experienced mysql/postgres admin with experience in master/slave setups can apply their knowledge to pet sets and get an outcome that is safe.  That means the problems they hit have to be obvious, and the failure modes also have to be obvious.  If another admin somewhere else runs a `kubectl drain`, they need to understand the impact they'll have on you (the person doing the drain may not be the mysql admin).  A set of useful tools and patterns will go a long way.

**Comment by bprashanth** [comment]
fencing in https://github.com/kubernetes/kubernetes/issues/31762

**Comment by tongpu** [comment]
@smarterclayton I think your described approach towards what Pet Sets should provide is a wise choice. In my opinion Kubernetes Pet Sets should just provide a framework for init/bootstrap containers to collect information about expected other members in the cluster, and all the heavy lifting should still be kept for the cluster frameworks to work out.

**Comment by smarterclayton** [comment]
Pet safety and storage safety proposal https://github.com/kubernetes/kubernetes/pull/34160

**Comment by erictune** [comment]
My take on beta:

Before Code Freeze:

- [ ] Naming
  - [x] pick name #27430
  - [ ] Rename Kind to StatefulSet and change version to v1beta1 (@chiachenk working on a PR)
  - [ ] fix kubectl describe to recognize statefulset instead of petset  ( @ymqytw working on a PR)
  - [ ] document cluster upgrade process
  - [ ] rename examples (@kow3ns working on a PR)
- [ ] Fencing
  - [ ] at-most-one-pod-per-index semantics is sufficient for beta ( #35145 being worked on by @foxish  ) 
  - [ ] e2e test that tests this semantic (part of  #35145 or follow on)

After Code Freeze but before 1.5:

- [ ] Dcoument and test procedure to convert resources from `PetSet/v1alpha1` to `StatefulSet/v1beta1` whilst upgrading from 1.4 to 1.5.
- [ ] Greatly expand documentation on best practices for PetSets, reasoning about availability in the face of node updates, debugging, etc, etc...
- [ ] Rename PetSet to StatefulSet in all kubernetes.github.io docs

Right after 1.5:
- [ ]  convert incubating Charts that use PetSet to StatefulSet and promote to stable if stable (@viglesiasce will lead this)

Soon after that:
- [ ] move filenames that contain `pet` to new names.
- [ ] more examples/Charts for more apps.

Everything else is for second beta or GA or not doing.

@smarterclayton

**Comment by bprashanth** [comment]
Theres something else we (well I) kind-of (read: completely) forgot. There's a useful feature we added to Services. The annotations is shown in many petset examples: https://github.com/kubernetes/kubernetes/blob/master/test/e2e/testing-manifests/petset/zookeeper/service.yaml#L6

It enables pets to make decisions about their peers being ready/unready using internal protocols, so eg: a http readiness probe timeout doesn't end up removing the DNS records for zookeeper-0, forcing a re-election. 

You can still create petsets without this annotation, of course, just don't give them a readiness probe. But that means no service can leverage the magic of a readiness probe (i.e there are 2 kinds of services, the governing Service, which always needs DNS, and any other type of overlay Service, which should ideally not direct traffic to an "unhealthy" pet). 

Thoughts on renaming this to `beta.services`? (https://github.com/kubernetes/kubernetes/pull/35713)

**Comment by chrislovecnm** [comment]
Guys this is a show stopper https://github.com/kubernetes/kubernetes/issues/33727 can we get it in by code freeze and back ported. We are about to go into prod with a bunch of pets, and seamless upgrades are super important to our sla.

**Comment by alexouzounis** [comment]
for us, being able to get the index/ordinal in an environment variable would be very useful. https://github.com/kubernetes/kubernetes/issues/30427 is tracking this and it would be great if it was included for 1.5

**Comment by smarterclayton** [comment]
The API changes are too significant and we're past the window for that (in
terms of review bandwidth).  It's just not going to make it for 1.5.

**Comment by erictune** [comment]
I've created a detailed documentation plan for PetSets, to be completed by 1.5 docs freeze: https://github.com/kubernetes/kubernetes.github.io/issues/1655

**Comment by dims** [comment]
Moving to 1.6 per @smarterclayton 's comment

**Comment by bprashanth** [comment]
I think Clayton was talking about https://github.com/kubernetes/kubernetes/issues/28718#issuecomment-257395432. PetSet should be in beta under the alias "StatefulSet" in 1.5. I'll leave it to the people actively working on the transition to decide when to close this, and what to migrate out into a "PetSet in GA" bug.

**Comment by Stono** [comment]
I need this in my Production life.

**Comment by dims** [comment]
@bprashanth @foxish Is it appropriate to move this to the next milestone or clear the 1.5 milestone? (and remove the non-release-blocker tag as well)

**Comment by bprashanth** [comment]
petset is beta in 1.5 and I think all issues discussed here have spin off bugs
